app_dart/lib/src/request_handling/authentication.dart

```diff
@@ -135,68 +139,99 @@ class AuthenticationProvider {
-            'id_token': idToken,
-          },
-        ));
-        final HttpClientResponse verifyTokenResponse = await verifyTokenRequest.close();
-
-        if (verifyTokenResponse.statusCode != HttpStatus.ok) {
-          final String body = await utf8.decodeStream(verifyTokenResponse);
-          log.warning('Token verification failed: ${verifyTokenResponse.statusCode}; $body');
-          throw const Unauthenticated('Invalid ID token');
-        }
-
-        final String tokenJson = await utf8.decodeStream(verifyTokenResponse);
-        TokenInfo token;
+    } else if (idTokenFromCookie != null || idTokenFromHeader != null) {
+      /// There are two possible sources for an id token:
+      ///
+      /// 1. Angular Dart app sends it as a Cookie
+      /// 2. Flutter app sends it as an HTTP header
+      ///
+      /// The Flutter application sends both since the client cookies are forwarded
+      /// in the request. One of these may be unauthenticated, so we need to try
+      /// authenticating both.
+      if (idTokenFromCookie != null) {
         try {
-          token = TokenInfo.fromJson(json.decode(tokenJson));
-        } on FormatException {
-          throw InternalServerError('Invalid JSON: "$tokenJson"');
-        }
+          return await authenticateIdToken(idTokenFromCookie,
+              clientContext: clientContext, log: log);
+        } catch (_) {}
+      }
 
-        final String clientId = await _config.oauthClientId;
-        assert(clientId != null);
-        if (token.audience != clientId) {
-          log.warning('Possible forged token: "${token.audience}" (expected "$clientId")');
-          throw const Unauthenticated('Invalid ID token');
-        }
+      if (idTokenFromHeader != null) {
+        return authenticateIdToken(idTokenFromHeader,
+            clientContext: clientContext, log: log);
+      }
+    }
 
-        if (token.hostedDomain != 'google.com') {
-          final bool isWhitelisted = await _isWhitelisted(token.email);
-          if (!isWhitelisted) {
-            throw Unauthenticated('${token.email} is not authorized to access the dashboard');
-          }
-        }
+    throw const Unauthenticated('User is not signed in');
+  }
+
+  @visibleForTesting
+  Future<AuthenticatedContext> authenticateIdToken(String idToken,
+      {ClientContext clientContext, Logging log}) async {
+    // Authenticate as a signed-in Google account via OAuth id token.
+    final HttpClient client = _httpClientProvider();
+    try {
+      final HttpClientRequest verifyTokenRequest =
+          await client.getUrl(Uri.https(
+        'oauth2.googleapis.com',
+        '/tokeninfo',
+        <String, String>{
+          'id_token': idToken,
+        },
+      ));
+      final HttpClientResponse verifyTokenResponse =
+          await verifyTokenRequest.close();
+
+      if (verifyTokenResponse.statusCode != HttpStatus.ok) {
+        final String body = await utf8.decodeStream(verifyTokenResponse);
```

Comment:
I'm skeptical, when the response isn't ok, should we decode stream?

Reply:
This code just shows added since it was moved to its own method.

The reason this decodes it is because the Google Auth API can give some reasoning to why the user was unauthenticated. Usually this is just "invalid token", but it can be other things like "token expired." This lets us better understand in the logs why authentication failures are happening.